File: Sources/Contracts/Moment/MomentContracts.swift
import Foundation

public enum MomentType: String, Codable, CaseIterable, Sendable {
    case prayer
    case scripture
    case sermon
    case event
    case creator
    case study
    case mission
    case thread
}

public enum TemporalState: String, Codable, CaseIterable, Sendable {
    case upcoming
    case live
    case recap
    case evergreen
}

public enum VerbFamily: String, Codable, CaseIterable, Sendable {
    case gather
    case deepen
    case invite
    case follow
}

public struct Moment: Codable, Equatable, Sendable {
    public let id: String
    public let type: MomentType
    public let temporalState: TemporalState
    public let refId: String
    public let ownerId: String
    public let createdAt: Int64

    public init(
        id: String,
        type: MomentType,
        temporalState: TemporalState,
        refId: String,
        ownerId: String,
        createdAt: Int64
    ) {
        self.id = id
        self.type = type
        self.temporalState = temporalState
        self.refId = refId
        self.ownerId = ownerId
        self.createdAt = createdAt
    }
}

public struct MomentFlags: Codable, Equatable, Sendable {
    public let momentSystemEnabled: Bool
    public let deepenActionsEnabled: Bool
    public let gatherLiveEnabled: Bool

    public static let off = MomentFlags(
        momentSystemEnabled: false,
        deepenActionsEnabled: false,
        gatherLiveEnabled: false
    )
}

public enum DeepenActionKind: String, Codable, CaseIterable, Sendable {
    case summarize
    case crossReference
    case generatePrayer
    case generateStudyGuide
    case generateDiscussion
    case generateDevotional
    case saveTo
}

public enum GatherActionKind: String, Codable, CaseIterable, Sendable {
    case prayLive
    case joinAudio
    case joinDiscussion
}

public enum ActionKind: Codable, Equatable, Sendable {
    case deepen(DeepenActionKind)
    case gather(GatherActionKind)
    case invite
    case follow
}

public struct MomentAction: Codable, Equatable, Sendable {
    public let id: ActionKind
    public let family: VerbFamily
    public let enabled: Bool
    public let reason: String?
}

public func availableActions(for moment: Moment, flags: MomentFlags) -> [MomentAction] {
    guard flags.momentSystemEnabled else {
        return []
    }

    var actions: [MomentAction] = []

    if flags.deepenActionsEnabled {
        actions.append(contentsOf: DeepenActionKind.allCases.map {
            MomentAction(id: .deepen($0), family: .deepen, enabled: true, reason: nil)
        })
    }

    if moment.temporalState == .live && flags.gatherLiveEnabled {
        actions.append(contentsOf: GatherActionKind.allCases.map {
            MomentAction(id: .gather($0), family: .gather, enabled: true, reason: nil)
        })
    }

    actions.append(MomentAction(id: .invite, family: .invite, enabled: false, reason: "uiDeferred"))
    actions.append(MomentAction(id: .follow, family: .follow, enabled: false, reason: "uiDeferred"))

    return actions
}

public enum BereanMode: String, Codable, CaseIterable, Sendable {
    case ask
    case discern
    case build
}

public enum SaveTarget: String, Codable, CaseIterable, Sendable {
    case prayerJournal
    case studyJournal
    case churchNotes
    case sermonCollection
    case savedTeachings
}

public struct DeepenRequest: Codable, Equatable, Sendable {
    public let moment: Moment
    public let action: DeepenActionKind
    public let requesterId: String
    public let bereanMode: BereanMode
    public let saveTarget: SaveTarget?
    public let locale: String?
}

public struct GuardianReview: Codable, Equatable, Sendable {
    public let passed: Bool
    public let policyVersion: String
    public let reason: String?
}

public struct DeepenResult: Codable, Equatable, Sendable {
    public let momentId: String
    public let action: DeepenActionKind
    public let output: String
    public let citations: [String]
    public let savedTo: SaveTarget?
    public let guardian: GuardianReview
    public let createdAt: Int64
}

public struct GatherRequest: Codable, Equatable, Sendable {
    public let moment: Moment
    public let action: GatherActionKind
    public let requesterId: String
}

public enum GatherStatus: String, Codable, Sendable {
    case gated
    case notImplemented
}

public enum GatherReason: String, Codable, Sendable {
    case complianceGateRequired
    case flagDisabled
    case v1StubOnly
}

public struct GatherResult: Codable, Equatable, Sendable {
    public let momentId: String
    public let action: GatherActionKind
    public let status: GatherStatus
    public let reason: GatherReason
}

public let momentFunctionRegion = "us-east1"

public let deepenFunctionNames = [
    "momentSummarize",
    "momentCrossReference",
    "momentGeneratePrayer",
    "momentGenerateStudyGuide",
    "momentGenerateDiscussion",
    "momentGenerateDevotional",
    "momentSaveTo",
]

public let gatherFunctionNames = [
    "momentPrayLive",
    "momentJoinAudio",
    "momentJoinDiscussion",
]

File: Contracts/moment/momentContracts.ts
export type MomentType =
    | "prayer"
    | "scripture"
    | "sermon"
    | "event"
    | "creator"
    | "study"
    | "mission"
    | "thread";

export type TemporalState = "upcoming" | "live" | "recap" | "evergreen";

export type VerbFamily = "gather" | "deepen" | "invite" | "follow";

export interface Moment {
    id: string;
    type: MomentType;
    temporalState: TemporalState;
    refId: string;
    ownerId: string;
    createdAt: number;
}

export interface Flags {
    moment_system_enabled: boolean;
    deepen_actions_enabled: boolean;
    gather_live_enabled: boolean;
    gather_compliance_gate_cleared: boolean;
}

export const defaultMomentFlags: Flags = {
    moment_system_enabled: false,
    deepen_actions_enabled: false,
    gather_live_enabled: false,
    gather_compliance_gate_cleared: false,
};

export type DeepenActionKind =
    | "summarize"
    | "crossReference"
    | "generatePrayer"
    | "generateStudyGuide"
    | "generateDiscussion"
    | "generateDevotional"
    | "saveTo";

export type GatherActionKind = "prayLive" | "joinAudio" | "joinDiscussion";

export type InviteActionKind = "invite";

export type FollowActionKind = "follow";

export type ActionKind =
    | DeepenActionKind
    | GatherActionKind
    | InviteActionKind
    | FollowActionKind;

export type ActionReason =
    | "complianceGateRequired"
    | "flagDisabled"
    | "uiDeferred"
    | "v1StubOnly";

export interface Action {
    id: ActionKind;
    family: VerbFamily;
    enabled: boolean;
    reason?: ActionReason;
}

export const v1DeepenEnabledMomentTypes: readonly MomentType[] = [
    "prayer",
    "scripture",
    "sermon",
    "event",
    "creator",
    "study",
    "mission",
    "thread",
];

export const v1GatherEligibleMomentTypes: readonly MomentType[] = [
    "prayer",
    "scripture",
    "sermon",
    "event",
    "creator",
    "study",
    "mission",
    "thread",
];

const deepenActions: readonly Action[] = [
    { id: "summarize", family: "deepen", enabled: true },
    { id: "crossReference", family: "deepen", enabled: true },
    { id: "generatePrayer", family: "deepen", enabled: true },
    { id: "generateStudyGuide", family: "deepen", enabled: true },
    { id: "generateDiscussion", family: "deepen", enabled: true },
    { id: "generateDevotional", family: "deepen", enabled: true },
    { id: "saveTo", family: "deepen", enabled: true },
];

const gatherActionKinds: readonly GatherActionKind[] = [
    "prayLive",
    "joinAudio",
    "joinDiscussion",
];

export function availableActions(moment: Moment, flags: Flags): Action[] {
    if (!flags.moment_system_enabled) {
        return [];
    }

    const actions: Action[] = [];

    if (flags.deepen_actions_enabled && v1DeepenEnabledMomentTypes.includes(moment.type)) {
        actions.push(...deepenActions);
    }

    const canSurfaceGather =
        moment.temporalState === "live" &&
        flags.gather_live_enabled &&
        v1GatherEligibleMomentTypes.includes(moment.type);

    if (canSurfaceGather) {
        actions.push(...gatherActionKinds.map((id): Action => ({
            id,
            family: "gather",
            enabled: flags.gather_compliance_gate_cleared,
            reason: flags.gather_compliance_gate_cleared ? undefined : "complianceGateRequired",
        })));
    }

    actions.push(
        { id: "invite", family: "invite", enabled: false, reason: "uiDeferred" },
        { id: "follow", family: "follow", enabled: false, reason: "uiDeferred" },
    );

    return actions;
}

export type BereanMode = "ask" | "discern" | "build";

export type SaveTarget =
    | "prayerJournal"
    | "studyJournal"
    | "churchNotes"
    | "sermonCollection"
    | "savedTeachings";

export interface DeepenRequest {
    moment: Moment;
    action: DeepenActionKind;
    requesterId: string;
    bereanMode: BereanMode;
    saveTarget?: SaveTarget;
    locale?: string;
}

export interface GuardianReview {
    passed: boolean;
    policyVersion: string;
    reason?: string;
}

export interface DeepenResult {
    momentId: string;
    action: DeepenActionKind;
    output: string;
    citations: string[];
    savedTo?: SaveTarget;
    guardian: GuardianReview;
    createdAt: number;
}

export interface GatherRequest {
    moment: Moment;
    action: GatherActionKind;
    requesterId: string;
}

export interface GatherResult {
    momentId: string;
    action: GatherActionKind;
    status: "gated" | "notImplemented";
    reason: "complianceGateRequired" | "flagDisabled" | "v1StubOnly";
}

export interface InviteRequest {
    moment: Moment;
    requesterId: string;
    recipientIds: string[];
}

export interface FollowRequest {
    moment: Moment;
    requesterId: string;
}

export const momentFunctionRegion = "us-east1";

export const deepenFunctionNames = [
    "momentSummarize",
    "momentCrossReference",
    "momentGeneratePrayer",
    "momentGenerateStudyGuide",
    "momentGenerateDiscussion",
    "momentGenerateDevotional",
    "momentSaveTo",
] as const;

export const gatherFunctionNames = [
    "momentPrayLive",
    "momentJoinAudio",
    "momentJoinDiscussion",
] as const;

export interface MomentFirestoreDocument extends Moment {
    schemaVersion: 1;
    updatedAt: number;
}

export interface MomentSaveDocument {
    schemaVersion: 1;
    momentId: string;
    ownerId: string;
    target: SaveTarget;
    deepenAction: DeepenActionKind;
    content: string;
    citations: string[];
    guardian: GuardianReview;
    createdAt: number;
}
